### ISSUE TITLE
parsing ObjectIDs

### ISSUE THREAD

**Issue by vorticalbox** [davidmdm/myzod#27]
I am using the myzod to validate user objects that are `{ _id: ObjectID, username: string }`  however there is no objectID type so i am using this schema.

```
const userSchema = mz.object({
  _id: mz.string().withPredicate(ObjectId.isValid, '_id must be an object ID'),
  username: mz.string(),
});
```

which works great for validating the input is an objectID the problem is that I want this to be stored as a ObjectID in the database and was wondering if there is nice way to parse this?

or a way to have a custom type 

```
const userSchema = mz.object({
  // pass in a custom typing so the interface correctly shows objectID rather than string
  _id: mz.custom(Types.Object).withPredicate(ObjectId.isValid, '_id must be an object ID'),
  username: mz.string(),
});
```

**Comment by davidmdm** [comment]
Hi @vorticalbox,

What is being passed in to your schema currently is a string or it wouldn't parse.
Some other types have coercion support from parsing a string representation to that type (numbers/dates/arrays).
There is no current way of taking a schema for a given type and returning a value of a different type when parsed. 

The suggestion is to wrap the schema yourself:

```typescript
const myParse = (input: unknown): MyType => {
  const data = schema.parse(input);
  return { ...data, _id: new ObjectID(data._id); }
}
```

and apply the transformation that way. 
Otherwise, I have considered adding a new `map` api to schemas in the past, which I might try to do as this is the second time I think I get this feature request.

Having a new api like this:

```typescript
const userSchema = mz.object({
  _id: mz.string()
           .withPredicate(ObjectId.isValid, '_id must be an object ID')
           .map(value => new ObjectID(value)),
  username: mz.string(),
});
```

**Comment by mscottnelson** [comment]
I am looking forward to when ts 4.1.0 lands especially for mapping capabilities like this, mostly because you gain the ability to *infer* a resultant object type based on incoming *string keys/values* without needing static string literals.

I guess you could say they are dynamic string literals? Generic string literals? Both seem somewhat like oxymorons...maybe "Composite string literal generics, usable as indexers." If you haven't started playing with them already, I highly recommend checking out the beta.

**Comment by vorticalbox** [comment]
> I am looking forward to when ts 4.1.0 lands especially for mapping capabilities like this, mostly because you gain the ability to _infer_ a resultant object type based on incoming _string keys/values_ without needing static string literals.
> 
> I guess you could say they are dynamic string literals? Generic string literals? Both seem somewhat like oxymorons...maybe "Composite string literal generics, usable as indexers." If you haven't started playing with them already, I highly recommend checking out the beta.

I will do!

**Comment by davidmdm** [comment]
@vorticalbox 
I implemented the `map` functionality as described above and used your example in the docs. This has been released in version 1.2.0.

Hence you can now have this (taken from readme):
```typescript
const ObjectIDSchema = myzod
                         .string()
                         .withPredicate(ObjectId.isValid, 'must be an object ID')
                         .map(value => new ObjectId(value));

// Infer<ObjectIDSchema> === ObjectId

const id = ObjectIDSchema.parse('507c7f79bcf86cd7994f6c0e');
id instanceof ObjectId; // true

const id2 = ObjectIDSchema.parse('some string'); // Throws VaidationError with message "must be an object ID"'
```

Hopefully this solves your issue! 
Closing the issue now, but would love to hear your feedback.
